packages/h5wasm/src/utils.ts

```diff
@@ -45,11 +45,13 @@ export function convertMetadataToDType(metadata: Metadata): DType {
   }
 
   if (isStringMetadata(metadata)) {
-    const { size: length, cset } = metadata;
+    const { size, cset, vlen } = metadata;
 
     return {
       class: DTypeClass.String,
-      length,
+      // For variable-length string datatypes, the returned value is the size of the pointer to the actual string and
+      // not the size of actual variable-length string data (https://portal.hdfgroup.org/display/HDF5/H5T_GET_SIZE)
+      length: vlen ? undefined : size,
```

Comment:
After the fix above, I was wondering why all my string datasets were interpreted as 4-character long. Then, I saw that they were in fact of variable-length, meaning that the `size` field did not reflect the actual string length.